src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,140 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
+public class Parser {
+
+    /**
+     * Parses user input to determine what action to take.
+     * @param storage A Storage instance managing a save file
+     * @param tasks A TaskList instance managing tasks
+     * @param ui A Ui instance managing displaying of output
+     */
+    public static void parse(Storage storage, TaskList tasks, Ui ui) {
+        boolean isRunning = true;
+
+        while (isRunning) {
+            try {
+                String input = ui.readCommand();
+                String[] inputSplit = input.split(" ", 2);
+
+                Command cmd = Command.getCommand(inputSplit[0]);
+
+                switch (cmd) {
+                case BYE:
+                    Ui.printMessage("Bye! Hope you visit again soon!");
+                    isRunning = false;
+                    break;
+                case FIND:
+                    if (inputSplit.length < 2) {
+                        throw new InsufficientArgumentsException("☹ OOPS!!! FIND requires something to search for.");
+                    } else {
+                        Ui.printMessage(tasks.findTask(inputSplit[1]));
+                    }
+                    break;
+                case LIST:
+                    tasks.displayTasks();
+                    break;
+                case DONE:
+                    if (inputSplit.length < 2) {
+                        throw new InsufficientArgumentsException("☹ OOPS!!! Missing command parameters!");
+                    } else {
+                        int index = Integer.valueOf(inputSplit[1]) - 1;
+                        if (index < 0 || index >= tasks.getTasks().size()) {
+                            throw new OutOfRangeException();
+                        } else {
+                            Task task = tasks.getTasks().get(index);
+                            task.markAsDone();
+                            Ui.printMessage("Nice! I've marked this task as done:\n\t" + task.toString());
+                            storage.storeTasks(tasks.getTasks());
+                        }
+                    }
+                    break;
+                case DELETE:
+                    if (inputSplit.length < 2) {
+                        throw new InsufficientArgumentsException("☹ OOPS!!! Missing command parameters!");
+                    } else {
+                        int index = Integer.valueOf(inputSplit[1]) - 1;
+                        if (index < 0 || index >= tasks.getTasks().size()) {
+                            throw new OutOfRangeException();
+                        } else {
+                            Task taskToDelete = tasks.getTasks().get(index);
```

Comment:
- a lot of arrowhead code here. try to refactor.
- I think many of the lines in this can be extracted as separate methods to improve SLAP